docs/interface_specification.md

```diff
@@ -12,28 +12,45 @@ BlackParrot is designed as a set of modular processor building blocks connected
-  - Typically this structure includes a pointer back to the original state that made the prediction; for example an index into a branch history table; or a field indicating whether the prediction came from BTB, BHT, or RAS (return address stack). All of these are private to the predictor and forwarded along by the BE.
-  - An example two-bit saturating counter branch prediction implementation, where high bit indicates the predicted direction, and the low bit indicates if the prediction is weak:
-    - bp\_fe\_branch\_metadata\_fwd\_s (only defined in FE)
-    - oldval: 2-bit state bits read from the branch history table (BHT)
-    - index: index of entry in branch history table
+
+    Invariant. Although the PC may be speculative, the value of _instruction_ is correct for that PC. We make similar assertions for the itlb -- the itlb mappings made by the FE are not speculative, and are required to be correct at all times.
+  - Branch prediction metadata (`bp_fe_branch_metadata_fwd_s`)
+
+    This structure is for the branch predictor to update its internal data structures based on feedback from the BE as to whether this particular PC/instruction pair was correct. The "forward packet" capturing this data is included in "attaboy" and "redirect" commands from the backend, which notify the frontend of a correct and incorrect prediction respectively.
+
+    The branch prediction metadata for our current frontend implementation contains:
+
+    - Flags indicating what type of jump triggered this prediction (branch, JAL or JALR) as well as whether it appeared to be a call or return
+    - Flags noting which internal branch prediction structures (BTB or RAS) were used to make this prediction
+    - The global history vector and BTB/BHT pointers from the time this prediction was made, to roll back to before this jump was predicted and undo any speculative updates
+
+    The BE does not look at the metadata, since this would violate our intended separation of the three components (BE,FE,ME) via a clean interface. The BE implementation must not be tightly coupled to the front end implementation. The bit pattern of the packet is blindly forwarded, and the above data is considered private to the frontend.
+
+    The backend knows that the frontend has made a incorrect prediction because it was sent an incorrect PC in the FE queue.
+
+    ---
+
+    An example two-bit saturating counter branch prediction implementation, where high bit indicates the predicted direction, and the low bit indicates if the prediction is weak:
```

Comment:
This bit feels somewhat out of place, and perhaps moreso now that I'm introducing details about what the BP frontend actually does in addition to just what the interface requires. IMHO making the additions I did is a good way to go, but I'm not sure what it means for this.

Reply:
I think it's fine to axe this, but add comment below (or paraphrase)

Reply:
I've added a paraphrasing of this idea after the paragraph you suggested below.